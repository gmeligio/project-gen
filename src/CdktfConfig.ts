import * as fs from 'fs';
import * as path from 'path';
import { Component, JsonFile, Project } from 'projen';
import { v4 as uuidv4 } from 'uuid';

export interface TerraformDependencyConstraint {
  /**
   * Name of the module / provider
   */
  readonly name: string;

  /**
   * Namespace of the module / provider
   */
  readonly namespace?: string;

  /**
   * Fully qualified name of the module / provider
   */
  readonly fqn?: string;

  /**
   * Path / url / registry identifier for the module / provider
   */
  readonly source?: string;

  /**
   * Version constraint (https://www.terraform.io/docs/language/providers/requirements.html#version-constraints)
   */
  readonly version?: string;
}

export type RequirementDefinition = string | TerraformDependencyConstraint;

export enum Language {
  TYPESCRIPT = 'typescript',
  PYTHON = 'python',
  CSHARP = 'csharp',
  JAVA = 'java',
  GO = 'go',
}

/**
 * Options for CdktfConfig
 */
export interface CdktfConfigOptions {
  /**
   * The command to run in order to synthesize the code to Terraform compatible JSON (language specific)
   */
  readonly app?: string;

  /**
   * Target language for building provider or module bindings. Currently supported: `typescript`, `python`, `java`, `csharp`, and `go`
   */
  readonly language?: Language;

  /**
   * Default: '.gen'. Path where generated provider bindings will be rendered to.
   *
   * @default ".gen"
   */
  readonly codeMakerOutput?: string;

  /**
   * Where the synthesized JSON should go. Also will be the working directory for CDKTF cli
   *
   * @default "cdktf.out"
   */
  readonly output?: string;

  /**
   * Unique identifier for the project used to differentiate projects
   *
   * @default "generated UUID"
   */
  readonly projectId?: string;

  /**
   * Whether to send crash reports to the CDKTF team
   *
   * @default "false"
   */
  readonly sendCrashReports?: string;

  /**
   * Terraform Providers to build
   */
  readonly terraformProviders?: RequirementDefinition[];

  /**
   * Terraform Modules to build
   */
  readonly terraformModules?: RequirementDefinition[];
}

export class CdktfConfig extends Component {
  /**
   * The command to run in order to synthesize the code to Terraform compatible JSON
   */
  public readonly json: JsonFile;

  /**
   * The command to run in order to synthesize the code to Terraform compatible JSON (language specific)
   */
  public readonly app: string;

  /**
   * Target language for building provider or module bindings. Currently supported: `typescript`, `python`, `java`, `csharp`, and `go`
   */
  public readonly language: string;

  /**
   * Unique identifier for the project used to differentiate projects
   */
  public readonly projectId: string;

  /**
   * Whether to send crash reports to the CDKTF team
   */
  public readonly sendCrashReports: string;

  /**
   * Terraform Providers to build
   */
  public readonly terraformProviders: RequirementDefinition[];

  /**
   * Terraform Modules to build
   */
  public readonly terraformModules: RequirementDefinition[];

  constructor(project: Project, options: CdktfConfigOptions) {
    super(project);

    if (!options.app) {
      throw new Error('Required option app is not specified.');
    }

    if (!options.language) {
      throw new Error('Required option language is not specified.');
    }

    const filename = 'cdktf.json';
    const projectIdKey = 'projectId';
    const filePath = path.join(project.outdir, filename);

    let projectId: string | undefined;
    if (fs.existsSync(filePath)) {
      const optionsFile = fs.readFileSync(filePath, 'utf8');
      const optionsObject = JSON.parse(optionsFile);

      projectId = optionsObject[projectIdKey];
    }

    this.app = options.app;
    this.language = options.language;
    this.projectId = projectId ?? uuidv4();
    this.sendCrashReports = options.sendCrashReports ?? 'false';
    this.terraformProviders = options.terraformProviders ?? [];
    this.terraformModules = options.terraformModules ?? [];

    const cdktfOptions = {
      app: this.app,
      language: this.language,
      projectId: this.projectId,
      sendCrashReports: this.sendCrashReports,
      terraformProviders: this.terraformProviders,
      terraformModules: this.terraformModules,
      ...options,
    };

    this.json = new JsonFile(project, filename, {
      readonly: true,
      obj: cdktfOptions,
    });
  }
}
