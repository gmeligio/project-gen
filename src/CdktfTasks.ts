import { Component, Project, Task } from 'projen';

/**
 * Adds standard AWS CDK tasks to your project.
 */
export class CdktfTasks extends Component {
  /**
   * Generate CDK Constructs for Terraform providers and modules.
   */
  public readonly get: Task;

  /**
   * Synthesizes Terraform code for the given app in a directory.
   */
  public readonly synth: Task;

  /**
   * Deploy the given stacks.
   */
  public readonly deploy: Task;

  /**
   * Destroy the given stacks.
   */
  public readonly destroy: Task;

  /**
   * Perform a diff (terraform plan) for the given stack.
   */
  public readonly diff: Task;

  constructor(project: Project) {
    super(project);

    this.get = project.addTask('get', {
      description: 'Generate CDK Constructs for Terraform providers and modules.',
      exec: 'cdktf get',
    });

    this.synth = project.addTask('synth', {
      description: 'Synthesizes Terraform code for the given app in a directory.',
      exec: 'cdktf synth',
    });

    this.deploy = project.addTask('deploy', {
      description: 'Deploy the given stacks.',
      exec: 'cdktf deploy',
      receiveArgs: true,
    });

    this.destroy = project.addTask('destroy', {
      description: 'Destroy the given stacks.',
      exec: 'cdktf destroy',
      receiveArgs: true,
    });

    this.diff = project.addTask('diff', {
      description: 'Perform a diff (terraform plan) for the given stack.',
      exec: 'cdktf diff',
    });
  }
}
