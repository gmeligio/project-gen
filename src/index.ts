export type { RequirementDefinition, TerraformDependencyConstraint } from './CdktfConfig';
export type { CdktfConfigOptions } from './CdktfConfig';
export { CdktfConfig, Language } from './CdktfConfig';
export { CdktfTasks } from './CdktfTasks';
export type { CdktfTypeScriptAppOptions } from './CdktfTypescriptApp';
export { CdktTypeScriptApp } from './CdktfTypescriptApp';
