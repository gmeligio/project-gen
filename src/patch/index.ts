export * from './GithubWorkflowPatch';
export * from './JsiiProjectPatch';
