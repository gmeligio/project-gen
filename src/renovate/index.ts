export * from './RenovateTransformer';
export * from './VersionTree';
