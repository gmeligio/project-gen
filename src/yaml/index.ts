export * from './YamlTransformer';
export * from './YamlTree';
