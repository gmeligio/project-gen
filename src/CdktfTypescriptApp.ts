import { typescript } from 'projen';
import { TypeScriptAppProject } from 'projen/lib/typescript';
import { CdktfConfig, CdktfConfigOptions, Language } from './CdktfConfig';
import { CdktfTasks } from './CdktfTasks';

export interface CdktfTypeScriptAppOptions extends typescript.TypeScriptProjectOptions {
  /**
   * The CDKTF cli configuration
   */
  readonly cdktfConfig?: CdktfConfigOptions;

  /**
   * Minimum target version this app can be run against
   *
   * @default "latest"
   */
  readonly cdktfVersion: string;

  /**
   * Construct version to use
   *
   * @default "latest"
   */
  readonly constructsVersion: string;

  /**
   * The file containing the CDKTF app
   *
   * @default "main.ts"
   */
  readonly cdktfAppFile?: string;
}

export class CdktfTypeScriptApp extends TypeScriptAppProject {
  /**
   * The CDKTF CLI options
   */
  public readonly cdktfConfig: CdktfConfig;

  /**
   * The tasks for this project
   */
  public readonly cdktfTasks: CdktfTasks;

  constructor(options: CdktfTypeScriptAppOptions) {
    if (!options.cdktfVersion) {
      throw new Error('Required field cdktfVersion is not specified.');
    }

    if (!options.constructsVersion) {
      throw new Error('Required field constructsVersion is not specified.');
    }

    if (options.cdktfConfig && options.cdktfConfig.language && options.cdktfConfig.language !== Language.TYPESCRIPT) {
      throw new Error(
        'TypeScript is the only supported language at this moment. The specified language must be Language.TYPESCRIPT.'
      );
    }

    super(options);

    this.addKeywords('cdktf');

    this.addDeps(`cdktf@${options.cdktfVersion}`, `constructs@${options.constructsVersion}`);

    this.addDevDeps(`cdktf-cli@${options.cdktfVersion}`);

    // No compile step because we do all of it in typescript directly
    this.compileTask.reset();

    const appFile = options.cdktfAppFile ?? 'main.ts';

    this.cdktfConfig = new CdktfConfig(this, {
      app: `npx ts-node ${appFile}`,
      language: Language.TYPESCRIPT,
      ...options.cdktfConfig,
    });

    this.cdktfTasks = new CdktfTasks(this);

    this.gitignore.exclude('.gen', 'cdktf.out');
  }
}
