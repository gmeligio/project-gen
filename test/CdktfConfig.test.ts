import * as fs from 'fs';
import { TestProject } from './TestProject';
import { CdktfConfig, Language, RequirementDefinition } from '../src/CdktfConfig';

jest.mock('fs', () => {
  return {
    // __esModule: true is important to solve error "TypeError: Cannot redefine property: existsSync"
    // https://github.com/aelbore/esbuild-jest/issues/26
    // https://stackoverflow.com/a/72885576/5405601
    __esModule: true,
    ...jest.requireActual('fs'),
  };
});

describe('CdktfConfig', () => {
  describe('cdktfVersion', () => {
    test('should throw an error if app is not specified', () => {
      const createCdktfConfig = () =>
        // @ts-ignore
        new CdktfConfig(new TestProject(), {
          language: Language.TYPESCRIPT,
        });
      expect(createCdktfConfig).toThrowError(new Error('Required option app is not specified.'));
    });

    test('should throw an error if language is not specified', () => {
      const createCdktfConfig = () =>
        // @ts-ignore
        new CdktfConfig(new TestProject(), {
          app: 'npx ts-node main.ts',
        });
      expect(createCdktfConfig).toThrowError(new Error('Required option language is not specified.'));
    });
  });

  describe('app', () => {
    test('can be overriden with custom command', () => {
      const command = 'npx ts-node --swc src/main.ts';

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: command,
        language: Language.TYPESCRIPT,
      });

      expect(cdktfConfig.app).toStrictEqual(command);
    });
  });

  describe('projectId', () => {
    afterEach(() => {
      jest.resetModules();
      jest.clearAllMocks();
    });

    test('should be `UUID`', () => {
      const command = 'npx ts-node --swc src/main.ts';

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: command,
        language: Language.TYPESCRIPT,
      });

      expect(cdktfConfig.projectId).toMatch(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/);
    });

    test('should reuse existing `projectId` when `cdktf.json` exists', () => {
      const existingProjectId = '12345678-1234-1234-1234-123456789012';

      // Mock the fs.existsSync function to return true for 'cdktf.json'
      jest.spyOn(fs, 'existsSync').mockImplementation((p: fs.PathLike) => {
        // cdktf.json path has the form /tmp/random-string/cdktf.json when testing because of how projen output is created when testing
        // Then we need to check if the path contains cdktf.json
        if (p.toString().includes('cdktf.json')) {
          return true;
        }

        return false;
      });

      // Mock the fs.readFileSync function to return a JSON with the existing projectId
      jest.spyOn(fs, 'readFileSync').mockImplementation((p: fs.PathOrFileDescriptor) => {
        // cdktf.json path has the form /tmp/random-string/cdktf.json when testing because of how projen output is created when testing
        // Then we need to check if the path contains cdktf.json
        if (p.toString().includes('cdktf.json')) {
          return JSON.stringify({
            projectId: '12345678-1234-1234-1234-123456789012',
          });
        }

        return '';
      });

      const cdktfConfig = new CdktfConfig(
        new TestProject({
          outdir: '.',
        }),
        {
          app: 'npx ts-node main.ts',
          language: Language.TYPESCRIPT,
        }
      );

      expect(cdktfConfig.projectId).toStrictEqual(existingProjectId);
    });
  });

  describe('sendCrashReports', () => {
    test('should be `"false"` by default', () => {
      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
      });

      expect(cdktfConfig.sendCrashReports).toStrictEqual('false');
    });

    test('can be overriden to true', () => {
      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        sendCrashReports: 'true',
      });

      expect(cdktfConfig.sendCrashReports).toStrictEqual('true');
    });
  });

  describe('terraformProviders', () => {
    test('should be empty by default', () => {
      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
      });

      expect(cdktfConfig.terraformProviders).toStrictEqual([]);
    });

    test('can be defined with name, source and version', () => {
      const terraformProviders: RequirementDefinition[] = [
        {
          name: 'aws',
          source: 'hashicorp/aws',
          version: '~> 3.22',
        },
      ];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformProviders,
      });

      expect(cdktfConfig.terraformProviders).toStrictEqual(terraformProviders);
    });

    test('can be defined with string', () => {
      const terraformProviders: RequirementDefinition[] = ['aws@~> 2.0'];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformProviders,
      });

      expect(cdktfConfig.terraformProviders).toStrictEqual(terraformProviders);
    });

    test('can be defined with name and version', () => {
      const terraformProviders: RequirementDefinition[] = [
        {
          name: 'aws',
          version: '~> 2.0',
        },
      ];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformProviders,
      });

      expect(cdktfConfig.terraformProviders).toStrictEqual(terraformProviders);
    });

    test('can contain objects and strings', () => {
      const terraformProviders: RequirementDefinition[] = [
        'aws@~> 2.0',
        {
          name: 'kubernetes',
          version: '~> 2.0',
        },
      ];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformProviders,
      });

      expect(cdktfConfig.terraformProviders).toStrictEqual(terraformProviders);
    });
  });

  describe('terraformModules', () => {
    test('should be empty by default', () => {
      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
      });

      expect(cdktfConfig.terraformModules).toStrictEqual([]);
    });

    test('can be defined with string', () => {
      const terraformModules: RequirementDefinition[] = ['terraform-aws-modules/vpc/aws@2.39.0'];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformModules,
      });

      expect(cdktfConfig.terraformModules).toStrictEqual(terraformModules);
    });

    test('can be defined with name and source', () => {
      const terraformModules: RequirementDefinition[] = [
        {
          name: 'local-module',
          source: './foo',
        },
      ];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformModules,
      });

      expect(cdktfConfig.terraformModules).toStrictEqual(terraformModules);
    });

    test('can be defined with name, source and version', () => {
      const terraformModules: RequirementDefinition[] = [
        {
          name: 'customAWSVpc',
          source: 'https://github.com/terraform-aws-modules/terraform-aws-vpc',
          version: '~> v2.0',
        },
      ];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformModules,
      });

      expect(cdktfConfig.terraformModules).toStrictEqual(terraformModules);
    });

    test('can contain objects and strings', () => {
      const terraformModules: RequirementDefinition[] = [
        'terraform-aws-modules/vpc/aws@2.39.0',
        {
          name: 'customAWSVpc',
          source: 'https://github.com/terraform-aws-modules/terraform-aws-vpc',
          version: '~> v2.0',
        },
      ];

      const cdktfConfig = new CdktfConfig(new TestProject(), {
        app: 'npx ts-node main.ts',
        language: Language.TYPESCRIPT,
        terraformModules,
      });

      expect(cdktfConfig.terraformModules).toStrictEqual(terraformModules);
    });
  });
});
