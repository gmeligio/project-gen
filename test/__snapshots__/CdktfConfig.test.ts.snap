// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdktfConfig generate cdktf.json following schema 1`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "app": "npx ts-node main.ts",
  "language": "typescript",
  "projectId": "c9e8d26b-bb1c-4788-a36a-2cdbc5d07833",
  "sendCrashReports": "false",
  "terraformModules": [],
  "terraformProviders": [],
}
`;
